pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/ExchangeNodeSplitterVisitor.java

```diff
@@ -0,0 +1,168 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.logical;
+
+import java.util.List;
+import org.apache.calcite.rel.RelDistribution;
+import org.apache.pinot.query.planner.partitioning.FieldSelectionKeySelector;
+import org.apache.pinot.query.planner.partitioning.KeySelector;
+import org.apache.pinot.query.planner.stage.AggregateNode;
+import org.apache.pinot.query.planner.stage.ExchangeNode;
+import org.apache.pinot.query.planner.stage.FilterNode;
+import org.apache.pinot.query.planner.stage.JoinNode;
+import org.apache.pinot.query.planner.stage.MailboxReceiveNode;
+import org.apache.pinot.query.planner.stage.MailboxSendNode;
+import org.apache.pinot.query.planner.stage.ProjectNode;
+import org.apache.pinot.query.planner.stage.SetOpNode;
+import org.apache.pinot.query.planner.stage.SortNode;
+import org.apache.pinot.query.planner.stage.StageNode;
+import org.apache.pinot.query.planner.stage.StageNodeVisitor;
+import org.apache.pinot.query.planner.stage.TableScanNode;
+import org.apache.pinot.query.planner.stage.ValueNode;
+import org.apache.pinot.query.planner.stage.WindowNode;
+
+
+public class ExchangeNodeSplitterVisitor implements StageNodeVisitor<StageNode, Void> {
```

Comment:
are we planning the make the exchange node splitter more generic going forward? cuz. 
1. if we were to support more complex logic such as pipeline breaker, then this API seems not enough to indicate that (as it only can split into MailboxSend/ReceiveNode; 
    - are we planning to support other type of split going forward? if so what would that be? such as subquery, etc)
2. if we decided to add another visitor to support other type of exchange that will not be converted into mailbox send/receive, then the logic should be merged with StageAssigner --> b/c the moment we do exchange split is the same moment we make decision to assign stage IDs. 

Let's make sure we are clear on what's the hierarch. us the example of subquery execution are we abstracting into:
1. convert exchange into subqueries and attach subqueryID
2. convert the rest of the exchanges into mailboxes + attach stageIDs; 
3. convert the rest of the exchanges into some other abstraction.

or
1. convert exchanges into subquery node and mailbox nodes (and some other nodes)
2. visit and assign subqueryID and stageID (and some other nodes)

Reply:
subPlanExchange --> indicate subPlan boundary --> it's own planNode root will not be run (held on broker) until all children returned and info is being converted into its own plan

Reply:
mailboxExchange --> indicate a planFragment boundary --> its own planRoot and all its children that will interact with it via mailbox; all planFragment will be dispatch from broker at the same time and will run on every server

Reply:
pipelineBreakerExchange --> indicate part of the planFragment can wait on another part of the fragment --> this wait decision is made on server

Reply:
Merged ExchangeNodeSplitterVisitor and StageIdAssigner to StageFragmenter which will look for *ExchangeNode* and split it to MailboxReceiveNode/MailboxSendNode pair.

Later on we can extend the logic here to introduce more types of *ExchangeNode*, like `SubPlanExchangeNode` or `PipelineBreakerExchangeNode`